numba/function.py

```diff
@@ -0,0 +1,204 @@
+    Signature instance as the argument, and returns an integer
+    representing the address or pointer value of a compiled function
+    with given signature.
+    """
+    if hasattr(func, '__wrapper_address__'):
+        # func can be any object that implements the
+        # __wrapper_address__ protocol.
+        addr = func.__wrapper_address__(sig)
+    elif isinstance(func, types.FunctionType):
+        cfunc = numba.cfunc(sig)(func)
+        addr = cfunc._wrapper_address
+    elif isinstance(func, CFunc):
+        if sig == func._sig:
+            addr = func._wrapper_address
+        else:
+            cfunc = numba.cfunc(sig)(func._pyfunc)
+            addr = cfunc._wrapper_address
+    else:
+        raise NotImplementedError(
+            f'get wrapper address of {type(func)} instance with {sig!r}')
+    if not isinstance(addr, int):
+        raise TypeError(
+            f'wrapper address must be integer, got {type(addr)} instance')
+    if addr <= 0:
+        raise ValueError(f'wrapper address of {type(func)} instance must be'
+                         f' positive integer but got {addr}')
+    return addr
+
+
+@unbox(FunctionType)
+def unbox_function_type(typ, obj, c):
+    sfunc = cgutils.create_struct_proxy(typ)(c.context, c.builder)
+
+    # Get obj wrapper address. The code below trusts that the function
+    # numba.function._get_wrapper_address exists and can be called
+    # with two arguments. However, if an exception is raised in the
+    # function, then it will be catched and propagated to the caller.
+    modname = c.context.insert_const_string(c.builder.module, 'numba.function')
+    numba_mod = c.pyapi.import_module_noblock(modname)
+    numba_func = c.pyapi.object_getattr_string(
+        numba_mod, '_get_wrapper_address')
+    c.pyapi.decref(numba_mod)
+
+    sig_obj = c.pyapi.unserialize(c.pyapi.serialize_object(typ.signature()))
+    addr = c.pyapi.call_function_objargs(numba_func, (obj, sig_obj))
+
+    with c.builder.if_else(cgutils.is_null(c.builder, addr),
+                           likely=False) as (then, orelse):
+        with then:
+            # propagate any errors in getting pyaddr to the caller
+            c.builder.ret(c.pyapi.get_null_object())
+        with orelse:
+            # _get_wrapper_address checks that pyaddr is int and
+            # nonzero, so no need to check it here. But it will be
+            # impossible to tell if the addr value actually
+            # corresponds to a memory location of a valid function.
+            sfunc.addr = c.pyapi.long_as_voidptr(addr)
+            c.pyapi.decref(addr)
+
+            # TODO: the following does not work on 32-bit systems
+            # TODO: is incref(obj) needed? where the corresponding
+            # decref should be called?
+            # TODO: see
+            # https://github.com/numba/numba/blob/77cb53ba8966a38bfbd3413559e4f04b12812535/numba/targets/boxing.py#L505
+            # as an alternative way for boxing
```

Comment:
Re TODO 1: this is obsolete, will remove.
Re TODO 2: this requires a test case that makes the current code fail. Whether such a test case exists, is not known atm.
Re TODO 3: the cited reference is not applicable (it would mean pickling Python functions but we need the address of the original function, not of the copy). Will remove.

I'll keep this conversation open until having more info re TODO 2.

Reply:
tested TODO 2 -> not an issue (pyaddr is always valid as python functions are attached to CFun instances).